### DISCUSSION TITLE
How to redirect a page with a countdown?

### DISCUSSION THREAD

**Discussion by kanbeq** [datenstrom/community#466] [Ask a question] [answered]
I would like to ask if possible to create an extension that will redirect the page to some other one
Eg. yellowsite.com/google -> google.com
Thanks in advance 😉

**Comment by schulle4u** [comment]
Datenstrom Yellow supports standard http301 redirecting out of the box using [page settings](https://extensions.datenstrom.se/help/markdown-cheat-sheet#settings). Here is an example content file: 

```
---
Title: Redirect to Google
Redirect: https://www.google.com
---
A page text usually is not needed, but in case the browser doesn't support the redirect, you can of course put [a link to Google.com here](https://www.google.com).
```

HTH
Steffen

**Comment by kanbeq** [comment]
I have read that page many times, but apparently I did not notice this option :wink: 
Can somehow be done so that this redirection is not immediate, but after 5 seconds, for example?

**Comment by schulle4u** [comment]
As far as I know this is not possible due to technical limitations of HTTP 301 redirects. You can use a meta refresh directive as fallback using the `onParsePageExtra()` function from the API to insert appropriate meta directives, however some browsers or browser extensions might block them for security reasons. I can throw together an extension for you later this day unless no smarter solution has been found. 😀

**Comment by kanbeq** [comment]
I found something else:
`
<meta http-equiv="Refresh" content="5; url=https://example.com">
`
This is probably what you meant with this, 

> meta refresh directive as fallback using the onParsePageExtra() function from the API to insert appropriate meta directives

but I could use a counter.
And at this point it would probably be necessary to do some extension.

**Comment by schulle4u** [comment]
Please download this file and put it into your `system/extensions` directory: 

https://github.com/schulle4u/yellow-extensions-schulle4u/raw/dev/refresh/refresh.php

In your page settings, use `Refresh: https://example.com` to create a meta refresh. The refresh time is set to 5 seconds by default, you can adjust it using `RefreshTime:` in your `system/settings/system.ini`. I can also add refresh time settings for individual pages if you require this. 

Let me know what you think. 😀

Steffen

**Comment by kanbeq** [comment]
I found a other solution, when you also tried to find a solution :laughing: :
I create a layout called "redirect.html" with the following content:

```
<?php $this->yellow->layout("header") ?>
<h3><center>Przekierowanie do <?php echo $this->yellow->page->getHtml("rTitle") ?> nastąpi za
<span id="countdown"><?php echo $this->yellow->page->getHtml("rTime") ?></span>s</center></h3>
<script type="text/javascript">
var seconds = <?php echo $this->yellow->page->getHtml("rTime") ?>;
function countdown() {
seconds = seconds - 1;
        if (seconds < 0) {
            // Chnage your redirection link here
            window.location = "<?php echo $this->yellow->page->getHtml("rURL") ?>";
        } else {
            // Update remaining seconds
            document.getElementById("countdown").innerHTML = seconds;
            // Count down using javascript
            window.setTimeout("countdown()", 1000);
        }
    }
countdown();
</script>
<?php $this->yellow->layout("footer") ?>
```
And I create a page at some address, with the following content:
```
---
Title: Google
Layout: redirect
rTitle: google.com
rURL: https://google.com/
rTime: 5
---
```

If I have solved my problem correctly, please close the topic.
Edit: Thanks for editing the bushes. A moment ago I was overwhelmed with the need to give these 3 stamps. I like your solution too, but I can set the redirection time only permanently and not depending on the site. And there is no countdown.

**Comment by schulle4u** [comment]
Edited the comment to correctly display the source code.  
I'm not a JS developer, but it looks like a working solution. However keep in mind that javascript highly depends on client support, and you always should create a fallback for e. g. text browsers and search engines. 

Feel free to close your own issues if everything is solved.

**Comment by schulle4u** [comment]
@kanbeq I've just updated and released my [refresh extension](https://github.com/schulle4u/yellow-extensions-schulle4u/tree/master/refresh). It takes some of your suggestions and provides an optional countdown for visitors. For the redirection part it uses a more robust meta refresh, allowing redirection if Javascript is disabled. Here is an example content file: 

```
---
Title: Redirecting to Google
Refresh: https://www.google.com
RefreshTime: 10
---
Redirecting in [refresh] seconds to https://www.google.comm. 
```

HTH
Steffen

**Comment by kanbeq** [comment]
Good. Almost like mine, but instead of a layout, a plugin. To be honest,
whether it's a plugin or a layout, it's practically one thing.
PS I would change the name from "refresh" to "redirect". I think you know why.

**Comment by BsNoSi** [comment]
The redirection by entry in markdown behaves different to a meta entry in a layout file. Why?

**Comment by schulle4u** [comment, marked as answer]
The `Redirect` setting in markdown manipulates the http header and sends a 301 status code, causing the client to immediately redirect to the desired location, while using a meta entry in the header layout is purely HTML-based and allows delayed redirections. 

BTW: My extension has been renamed as suggested and is [available here](https://github.com/schulle4u/yellow-extensions-schulle4u/tree/main/redirect). Latest core required.
